src/app/services/store/store.service.ts

```diff
@@ -0,0 +1,48 @@
+import { Observable, Subject, BehaviorSubject } from 'rxjs';
+import {  map } from 'rxjs/operators';
+
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http'; 
+
+import { Product } from '../../models/product.model';
+import { CartProduct } from 'src/app/models/cart-product.model';
+
+@Injectable()
+export class StoreService {
+
+  private products$: Observable<Product[]>;
+
+  constructor(private http: HttpClient) {
+    this.products$ = this.getJSON();
+  }
+
+  getProducts(): Observable<Product[]> {
+    return this.products$;
+  }
+
+  getProductByName(name: string): Observable<Product> {
+    return this.products$.pipe(
+      map((products: Product[]) => products.find((product: Product) => product.name === name))
+    );
+  }
+
+  purchaseProducts(cart: CartProduct[]) {
+    this.products$ = this.products$.pipe(map((products: Product[]) => {
```

Comment:
is it even working?? map do not emits anything to the observable stream.... when u buy product u can c the out of stock message?  u should save your products in behaviour subject

Reply:
i did see but i switched anyway